enhancements/oc/olm-mirroring.md

```diff
@@ -0,0 +1,121 @@
+- `--auth-token=string` - the auth token ([instructions](https://github.com/operator-framework/operator-courier#authentication)) for authenticating with appregistry.
+- `--appregistry-endpoint=url` - the CNR endpoint to authenticate against. Defaults to `"https://quay.io/cnr"`, the endpoint used by OpenShift 4.1-4.3.
+- `--appregistry-namespace=string` - the namespace in appregistry to mirror. Each repository within the namespace represents one operator. 
+- `--db-only` - if true, and `--save` is specified, the operator database will be saved in the directory noted by `--save`.
+
+This command generates and mirrors in one step, because we do not assume that there is any registry available aside from the target disconnected registry. `--save` can be used to save the image locally without a registry available.
+
+### Extract the contents of a catalog for mirroring
+
+`oc` will be extended with a second command:
+
+```sh
+$ oc adm catalog mirror --from=quay.io/ecordell/community-operators-catalog:4.2.1 --to=localhost:5000/community-operators --to-manifests=./mirror-manifests
+mirroring ...
+
+$ ls ./mirror-manifests
+imagecontentsourcepolicy.yaml
+catalogsource.yaml
+```
+
+`oc adm catalog mirror` will:
+
+- Pull the catalog image referenced by `--from`, using `oc` machinery
+- Read the database to get the list of operator and operand images
+- Build a mapping of images to the disconnected registry
+- Mirror all of the referenced images by (interally) calling `oc image mirror`
+- Output a set of manifests that, if applied to a cluster that has access to the mirrored images, will correctly configure nodes and OLM to use those images.
+
+It has the following flags:
+
+- `--from=ref` - a catalog image (such as one built by `oc catalog build`)
+- `--from-archive=path` - a reference to a tarball image, such as one built with `oc adm catalog build --save=path`
+- `--to=ref` - the location that the image will be mirrored. This should include only a registry and a namespace that images can be mirrored into.
+- `--to-manifests=path` - default `./manifests`, the path at which manifests required to mirror these images will be created. This includes an `ImageContentSourcePolicy` that can configure nodes to translate between the image references stored in operator manifests and the mirrored registry, and a `CatalogSource` that configures OLM to read from the mirrored catalog image referenced by `--from`.
+
+### Full Example
+
+```sh
+$ oc adm catalog build --appregistry-namespace=community-operators --to disconnected-registry:5000/openshift/community-operators-catalog:4.2.1 
+pulling quay.io/community-operators/etcd...
+buliding catalog...
+mirroring catalog...
+quay.io/community-operators mirrored to disconnected-registry:5000/openshift/community-operators-catalog:4.2.1
+
+$ oc catalog mirror --from=disconnected-registry:5000/openshift/community-operators-catalog:4.2.1 --to=disconnected-registry:5000/community-operators
```

Comment:
I would consider making these positional arguments, which is common for "copy"-type commands. Some examples pulled from help text:

`cp [OPTION]... [-T] SOURCE DEST`

`rsync [OPTION]... SRC [SRC]... DEST`

`skopeo copy [command options] SOURCE-IMAGE DESTINATION-IMAGE`

Reply:
> Usage:
  oc image mirror SRC DST [DST ...] [flags]

I think that we should conform to the mirror command that already exists

Reply:
That seems reasonable -

Do you think it's clear that, in this case, the SRC is _not_ copied, and that DEST is a _namespace_ to put a large set of images into? That would be my only concern in making this parallel the more "traditional" copying APIs.

Reply:
I think it's clear, and matches the idea of copying several files into a directory. You might consider putting a trailing slash in your examples to make that concept especially clear.

Reply:
Also, I don't think `--to=` does anything more or less to clarify the nature of the destination.

Reply:
I would prefer you remain consistent with the existing image mirror command vs new syntax, unless it's truly necessary.

Reply:
Updated